testing/multi-env-test-engine/src/main/java/org/projectnessie/junit/engine/MultiEnvTestFilter.java

```diff
@@ -60,14 +62,17 @@ private FilterResult filter(Class<?> testClass, UniqueId id) {
       }
     } else {
       // check whether any of the extensions declared by the test recognize the version segment
-      boolean matched =
+      Set<String> registeredMultiEnvSegmentTypes =
           registry.stream(testClass)
-              .anyMatch(
-                  ext ->
-                      id.getSegments().stream()
-                          .anyMatch(s -> ext.segmentType().equals(s.getType())));
+              .map(MultiEnvTestExtension::segmentType)
+              .collect(Collectors.toUnmodifiableSet());
+
+      Stream<String> segmentTypesInTestId = id.getSegments().stream().map(Segment::getType);
 
-      if (matched) {
+      boolean atLeastOneSegmentTypeIsMultiEnv =
+          segmentTypesInTestId.anyMatch(registeredMultiEnvSegmentTypes::contains);
+
+      if (atLeastOneSegmentTypeIsMultiEnv) {
         return FilterResult.included(null);
```

Comment:
If we have two (unrelated) tests in the same suite: A with X1 and X2 environments and B with X1, X2, Y1, Y2, Y2, A and B will each run 6 times, while A should run 2 times and B 6 times... Did I misinterpret perhaps? The existing `CartesianProductTest` is too simple, I'm afraid, it does not account for cross-test influences.

Basically, a test should not be attached to segments that come from extensions, which it does not declare, but that happens now I guess (I did not debug, so I may be mistaken).

I hope impl. would be simpler if segment types were declared on class files (testClass -> extensionClass -> segmentType value). We only need extension instances for discovering segment values, but we can do that once at JUnit5 discovery time, and then make the test descriptors and filtering base only on class data. WDYT?